unblock mom on execjob_end event

riyazhakki opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* *Currently, pbs_mom blocks upon execution of an execjob_end hook. In some cases the hook could take a considerable amount of time to complete. For example, running Cray's node health check. In such cases it would be better if the hook was non-blocking.*

#### Affected Platform(s)
* *Linux All*

#### Cause / Analysis / Design
* *pbs_mom loop through the execjob_end hook scripts and blocks until it's child process finishes the execution of the execjob_end hook script.*

#### Solution Description
* *Now execjob_end script will start execution in the background, once the child process exits from execution of hook script, it process result and creates a task for the pbs_mom to look for further scripts execution. If no hook script to run, then replies the outstanding batch request.*

#### Testing logs/output
* *Please attach your test log output from running the test you added (or from existing tests that cover your changes)*
[PTL_test_after_change.txt](https://github.com/PBSPro/pbspro/files/2666750/PTL_test_after_change.txt)
[PTL_test_before_change.txt](https://github.com/PBSPro/pbspro/files/2666751/PTL_test_before_change.txt)


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

prakashcv13 commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,300 @@ record_job_last_hook_executed(unsigned int hook_event,
+
+	run_hook(phook, phook->event, hook_input,
+			PBS_MOM_SERVICE_NAME, mom_host, 0, post_execjob_end_hook,
+			hook_infile, hook_outfile, hook_datafile, MAXPATHLEN+1  );
```
extra spaces before )

riyazhakki replied:
Done

saimanikant commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,131 @@
+                           n=100, max_attempts=10, interval=2)
+        self.mom.log_match("Job;%s;executed execjob_end hook" % jid2,
+                           n=100, max_attempts=10, interval=2)
+        time.sleep(20)
```
Sometimes the hook execution can overlap with ptl, trying to restart the mom. Hence this sleep buffer was added. Changed this with another mom log match.
Thanks

saimanikant commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,131 @@
+        self.mom.log_match("Job;%s;executed execjob_end hook" % jid,
+                           n=100, max_attempts=10, interval=2)
+        self.mom.log_match("Job;%s;Job is rejected" % jid,
+                           n=100, max_attempts=10, interval=2)
```
This test case is to ensure we are retaining the existing behavior

anamikau replied:
@saimanikant I agree with Prakash. This test needs to be extended to test multiple end job hooks do not block each other.

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,300 @@ record_job_last_hook_executed(unsigned int hook_event,
+			log_event(log_type, log_class, LOG_ERR, log_id, log_buffer);
+			record_job_last_hook_executed(phook->event, phook->hook_name, pjob, hook_outfile);
+			break;
+		default:
```
I have now removed check for non zero exit and the same moved to the default case so case 0 is required.

riyazhakki commented on src/resmom/requests.c:
```diff
@@ -619,23 +620,18 @@ req_deletejob(struct batch_request *preq)
 	 * prevent sending more OBIT messages
 	 */
+	pjob->ji_preq = preq;
+	hook_input = (mom_hook_input_t *) malloc (sizeof(mom_hook_input_t));
```
Done

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,300 @@ record_job_last_hook_executed(unsigned int hook_event,
+	hook			*phook = NULL;
+	mom_hook_input_t *hook_input = (mom_hook_input_t *) ptask->wt_parm1;
+
+	if (ptask->wt_parm2 == NULL){
```
Done

bayucan commented on src/resmom/mom_hook_func.c:
```diff
@@ -3192,6 +3474,28 @@ mom_process_hooks(unsigned int hook_event, char *req_user, char *req_host,
+				pjob->ji_qs.ji_un.ji_momt.ji_exitstat;
+			pjob->ji_wattr[(int)JOB_ATR_exit_status].at_flags |= (ATR_VFLAG_SET | ATR_VFLAG_MODCACHE);
+
+			pjob->ji_execjob_end_hook_event_started = 1;
```
I see pjob->ji_execjob_end_hook_event_started set to 1, but where in the code does this ever get set to 0?  Also, there should be an initialization entry for this field in src/server/job_func.c:job_alloc().

riyazhakki replied:
Thanks for the review Al, 
Done, I initialized in job_alloc (),  a reset is not required as the job will get purged after this event.

prakashcv13 replied:
it would be prudent to reset this field once all the execjob_end hooks are executed.

bayucan commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,285 @@ record_job_last_hook_executed(unsigned int hook_event,
+	char   reboot_cmd[HOOK_BUF_SIZE]  = {'\0'};
+	char   hook_outfile[MAXPATHLEN+1] = {'\0'};
+	pid_t  mypid = ptask->wt_event;
+	hook   *phook	= (hook *)ptask->wt_parm1;
```
Check if 'ptask' is not NULL before referencing it.

riyazhakki replied:
NULL check for this pointer is done after malloc in set_task function.

riyazhakki replied:
Done, I added a check NULL check for the ptask.

bayucan commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,285 @@ record_job_last_hook_executed(unsigned int hook_event,
+	char   hook_outfile[MAXPATHLEN+1] = {'\0'};
+	pid_t  mypid = ptask->wt_event;
+	hook   *phook	= (hook *)ptask->wt_parm1;
+	mom_hook_input_t *hook_input 	= (mom_hook_input_t *) ptask->wt_parm2;
```
same here, check if ptask is not NULL first.

riyazhakki replied:
Done

bayucan commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,285 @@ record_job_last_hook_executed(unsigned int hook_event,
+	pid_t  mypid = ptask->wt_event;
+	hook   *phook	= (hook *)ptask->wt_parm1;
+	mom_hook_input_t *hook_input 	= (mom_hook_input_t *) ptask->wt_parm2;
+	job *pjob	= (job *) hook_input->pjob;
```
Check if hook_input obtained is not NULL.

riyazhakki replied:
Done.

bayucan commented on src/resmom/requests.c:
```diff
@@ -548,6 +548,8 @@ return_file(job *pjob, enum job_file which, int sock)
 		return (-1);
 }
 
+#define HOOK_RUNNING_IN_BACKGROUD (3)
```
I think you meant *BACKGROUND and not *BACKGROUD.
Also, this define should be in an include file that is seen by mom_hook_func.c where inside mom_process_hooks(), instead of returning the hard-wired 3, it should be returning this macro instead.

riyazhakki replied:
Done.

prakashcv13 replied:
as we have defined HOOK_RUNNING_IN_BACKGROUND in the header file, we can remove this #define.

riyazhakki replied:
Done

bayucan commented on src/resmom/mom_hook_func.c:
```diff
@@ -3192,6 +3474,28 @@ mom_process_hooks(unsigned int hook_event, char *req_user, char *req_host,
+			pjob->ji_wattr[(int)JOB_ATR_exit_status].at_flags |= (ATR_VFLAG_SET | ATR_VFLAG_MODCACHE);
+
+			pjob->ji_execjob_end_hook_event_started = 1;
+			return (3); /* hook script started in child process*/
```
This should return the HOOK_RUNNING_IN_BACKGROUND...

riyazhakki replied:
Done.

saimanikant commented:
[Ptl logs.txt](https://github.com/PBSPro/pbspro/files/2679653/Ptl.logs.txt)

prakashcv13 commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -76,14 +76,12 @@ def test_execjob_end_non_blocking(self):
             date_time1, '%m/%d/%Y %H:%M:%S')))
-        (_, str1) = self.mom.log_match(
-                                       "executed exechost_periodic hook",
+        (_, str1) = self.mom.log_match("executed exechost_periodic hook",
```
@saimanikant - let us display the log times to be sure so that the sequence is evident in the test output.

anamikau replied:
@prakashcv13 displaying the time is good but how is it going to help if those timings are not right. test should have a check in itself that date_time2 < date_time3. or use starttime and endtime attributes of log_match to make sure periodic hook related messages appears between them.

anamikau replied:
Ah yes missed to see that. Either update the interval=1 to make sure this message appears within 10s or add an end time.

prakashcv13 replied:
consider using convert_date_time() from logutils here and in other places in the test

saimanikant commented:
[Ptl logs.txt](https://github.com/PBSPro/pbspro/files/2684737/Ptl.logs.txt)

prakashcv13 commented on src/resmom/requests.c:
```diff
@@ -619,23 +622,25 @@ req_deletejob(struct batch_request *preq)
-	hook_output.fail_action = &hook_fail_action;
+	if (mom_process_hooks(HOOK_EVENT_EXECJOB_END,
+		PBS_MOM_SERVICE_NAME, mom_host, hook_input,
+		hook_output, hook_msg, sizeof(hook_msg), 1) == HOOK_RUNNING_IN_BACKGROUD)
```
this should be HOOK_RUNNING_IN_BACKGROUND.

riyazhakki replied:
Done

Klovely commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,150 @@
+        self.mom.log_match("Job;%s;executed %s hook" % (jid, hook_name2),
+                           n=100, max_attempts=10, interval=2, existence=False)
+
+    def test_execjob_end_multi_job(self):
```
Could you please add one more scenario using  multi-node job (job which runs on both mother superior and sister mom )

riyazhakki replied:
We have a multi-node job test  [here](https://github.com/PBSPro/pbspro/blob/master/test/tests/resilience/pbs_hook_alarm_large_multinode_job.py#L181).

Klovely replied:
@Riyaz and @saimanikant as we discussed offline, the test you mention here is not fulfilling the objective of our use case,  it using 
self.server.create_vnodes(self.mom.shortname, a, 5000, self.mom)
and vnode is created with one mom and job submitted in this test case is include multi vnode not sister mom.
Our objective is to cover test and confirm the scenario when execjob_end should not block  sister mom when it running on sister mom as well .

anamikau replied:
@saimanikant the test you are pointing is not checking any blocking calls when execjob_end hook is run on remote moms. you can add a similar test as "test_execjob_end_non_blocking" with remote moms.

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,302 @@ record_job_last_hook_executed(unsigned int hook_event,
+/**
+ * @brief
+ * This function runs after execution of a single execjob_end hook and
+ * process the results of hook, on successful hook execution new task
```
Done

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,302 @@ record_job_last_hook_executed(unsigned int hook_event,
+ * process the results of hook, on successful hook execution new task
+ * will be created to run the next hook script. If there was  an
+ * error (the hook process returned a non-zero exit status) it does
+ * not create the new task for the next hook script and sends replies
```
Done

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,302 @@ record_job_last_hook_executed(unsigned int hook_event,
+	job				*pjob = NULL;
+	pbs_list_head	*head_ptr = NULL;
+	hook			*phook = NULL;
+	mom_hook_input_t *hook_input = (mom_hook_input_t *) ptask->wt_parm1;
```
Done

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3192,6 +3492,28 @@ mom_process_hooks(unsigned int hook_event, char *req_user, char *req_host,
+	if (hook_event == HOOK_EVENT_EXECJOB_END) {
+		/* Unblock MOM on execjob_end event */
+
+		struct 			work_task task;
```
Removed

riyazhakki commented on src/resmom/requests.c:
```diff
@@ -619,23 +622,25 @@ req_deletejob(struct batch_request *preq)
+			 * request will be taken care of in run_execjob_end_hooks
+			 * function 
+			 */
+			return;
```
I think we send IS_DISCARD_JOB to the running job, in our case job has been completed and we have received the exit status.

riyazhakki commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,150 @@
+        """
+        Test to make sure pbs_mom is not blocked upon
+        execution of an execjob_end hook.
+        """
```
Changed

riyazhakki commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,150 @@
+    def test_execjob_end_hook_order_and_reject(self):
+        """
+        Test with mutiple execjob_end hooks with with
+        different order for a job with one of the hooks rejecting the job.
```
Done

riyazhakki commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,150 @@
+        j = Job(TEST_USER)
+        j.set_sleep_time(1)
+        jid1 = self.server.submit(j)
+        j.set_sleep_time(1)
```
Done

riyazhakki commented:
> it would be prudent to reset this field once all the execjob_end hooks are executed.

I could not find the right place to reset it because after this hook event the job will get purged.

riyazhakki commented:
[ptl_log.txt](https://github.com/PBSPro/pbspro/files/2686974/ptl_log.txt)
Attaching latest test log

riyazhakki commented:
[ptl_log.txt](https://github.com/PBSPro/pbspro/files/2687424/ptl_log.txt)

anamikau commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,162 @@
+        j = Job(TEST_USER)
+        j.set_sleep_time(1)
+        jid1 = self.server.submit(j)
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
```
This will cause a race condition since 1s is too short. remove this check.

saimanikant replied:
This check is required since we want job 2 to be submitted after job1 has started.  I'll increase the sleep time of job1 to 3s Instead to avoid race condition.

anamikau replied:
@saimanikant 3s is also not enough. Either you put 10s or add a sleep of 2s between job submissions or make sure job1 is finished before submitting job2.

anamikau commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,162 @@
+        epoch1 = int(time.mktime(time.strptime(
+            date_time1, '%m/%d/%Y %H:%M:%S')))
+        (_, str1) = self.mom.log_match("Job;%s;executed execjob_end hook" %
+                                       jid2, starttime=epoch1, n=100,
```
we should also add an endtime of epoch1+10 (10 because we are sleeping for 10s in the end job hook) since we really expect this to appear before that. it is quite possible there could be some delay in 2 jobs execution causing second job message to be printed little late.

anamikau replied:
@saimanikant max_attempts=10 interval=2 will wait for 20s. in that case you might want to update interval=10.
Also please add comments in the code explaining what you are trying to verify.

anamikau replied:
@saimanikant I just realized that max_attempts will only check now+10s. it is quite possible that now is not equals to epoch1 but few seconds away. Hence I suggest to update to endtime=epoch1+10. Also increase the hook sleep time to 15s to avoid delays specially hook copy to moms.

saimanikant replied:
Agreed! I'll update this to endtime=epoch1+10. Hook execution starts only after the hook is copied to mom. So I think there's no need to increase sleep time.

anamikau commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -160,3 +163,52 @@ def test_execjob_end_multi_job(self):
                                        max_attempts=10, interval=2)
```
endjob hook for job1 should end within 10s when it started. I do not understand why are you checking it against the start time of job2's end job hook?

saimanikant replied:
since we are trying to ensure execjob_end log message of jid2 is logged between the execjob_end started for jid1 log message and execjob_end execution ended fir jid1 log message.

anamikau commented:
Also please provide the latest execution logs

saimanikant commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -132,14 +133,15 @@ def test_execjob_end_hook_order_and_reject(self):
         j.set_sleep_time(1)
         jid1 = self.server.submit(j)
-        self.server.expect(JOB, {'job_state': 'R'}, id=jid1)
+        self.server.expect(JOB, {'job_state': 'E'}, id=jid1)
```
We need jid1 to be in E state for submitting jid2, and jid1 will be in E state for at least 10 secs because of sleep in execjob_end hook.

anamikau replied:
Ah I missed that .. thanks @saimanikant for explaining it offline.

saimanikant commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -160,3 +163,52 @@ def test_execjob_end_multi_job(self):
+            epoch2 = int(time.mktime(time.strptime(
+                date_time2, '%m/%d/%Y %H:%M:%S')))
+            (_, str1) = mom.log_match(
+                "Job;%s;execjob_end hook ended" %
```
Since we are trying to ensure hook message for periodic hook is logged between the 2 execjob_end hook log messages.

anamikau replied:
Thanks for explaining this offline @saimanikant

saimanikant commented:
[Ptl logs.txt](https://github.com/PBSPro/pbspro/files/2702414/Ptl.logs.txt)

arungrover commented on src/resmom/requests.c:
```diff
@@ -619,23 +620,27 @@ req_deletejob(struct batch_request *preq)
-	hook_output.reject_errcode = &reject_errcode;
-	hook_output.last_phook = &last_phook;
-	hook_output.fail_action = &hook_fail_action;
+	if (mom_process_hooks(HOOK_EVENT_EXECJOB_END,
```
I think a similar change is needed in is_request() function as well (src/resmom/mom_server.c)

riyazhakki replied:
Done

arungrover commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,313 @@ record_job_last_hook_executed(unsigned int hook_event,
+	}
+
+	/* Create task to check and run next hook script */
+	new_task = set_task(WORK_Immed, 0,
```
I think we are leaking new_task here. I don't see run_execjob_end_hooks freeing it. Also, if you choose to free it in the callback function then you will have to change mom_process_hook as well because there you are passing stack variable.

riyazhakki replied:
No, we are not leaking new_task here, this will free in dispatch_task function.

arungrover commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,313 @@ record_job_last_hook_executed(unsigned int hook_event,
+		return;
+	}
+
+	hook_input = (mom_hook_input_t *) ptask->wt_parm2;
```
I don't see run_hook() setting wt_parm2, All I see is that run_hook() is setting hook pointer in wt_parm1. So wouldn't this always fail?

riyazhakki replied:
I am setting wt_param2 in run_hook() at [#L805](https://github.com/PBSPro/pbspro/blob/f2da3a05753ca04e8577528ac742180a7cd9ebd3/src/resmom/mom_hook_func.c#L805).

arungrover commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,215 @@
+            jid, starttime=epoch2-1, n=100,
+            max_attempts=10, interval=2)
+        date_time3 = str1.split(";")[0]
+        self.logger.info(
```
I think you should assert on the time difference between date_time1 and date_time2 to be 3 seconds or less (or maybe 5 seconds for slower machines but definitely less than 10)

saimanikant replied:
We are capturing date_time because we need to verify hook messages are logged in the order mentioned, to confirm mom is not blocked on execjob_end hook execution. The order is verified with the use of starttime and endtime parameters.

saimanikant replied:
In case of failure to find hook's log message in the given time period (starttime=epoch1-1 and endtime=epoch1+10) then test will fail with log message not found error. Only in case of success, we'll see the final log message. There's no need of asserting the time difference because starttime and endtime parameters in this test will ensure that hook's log message is matched in this interval.

riyazhakki commented:
Hi Arun, 
Thanks for reviewing the code. I will address your comments in the next commit.

visheshh commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,313 @@ record_job_last_hook_executed(unsigned int hook_event,
+post_execjob_end_hook(struct work_task *ptask)
+{
+
+	int    accept_flag = 1;
```
use unsigned short for flags

riyazhakki commented:
[ptl_log.txt](https://github.com/PBSPro/pbspro/files/2718482/ptl_log.txt)

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3077,313 @@ record_job_last_hook_executed(unsigned int hook_event,
+			snprintf(log_buffer, LOG_BUF_SIZE-1,
+				"%s hook '%s' encountered an exception, "
+				"request rejected",
+				hook_event_as_string(phook->event), phook->hook_name);
```
This is to make the statement more generic to all hook events, in case we decide to unblock the MOM on another hook event too.

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3192,6 +3503,27 @@ mom_process_hooks(unsigned int hook_event, char *req_user, char *req_host,
 		default:
 			return (-1); /* unexpected event encountered */
 	}
+#ifndef WIN32
```
This change is required for Linux platform only.

riyazhakki commented:
[valg.txt](https://github.com/PBSPro/pbspro/files/2719942/valg.txt)
Attaching valgrind logs

subhasisb commented on src/resmom/catch_child.c:
```diff
@@ -1253,6 +1253,15 @@ send_obit(job *pjob, int exval)
 	struct resc_used_update rud;
 	pbs_list_head vnl_changes;
 
+	if (pjob->ji_execjob_end_hook_event_started)
```
I think the code is becoming execjob_end hook specific. We would eventually make all the job lifecyle event hooks non-blocking for the mom, and thus the code is best if it stays generic. Please try to avoid hook event specific variables, and code structures. How about using something like "some job lifecyle events are being processed" flag, instead of a specific one. The other way to think about this to think of what "other" changes you would want to do if you had to make the execjob_launch hook non-blocking for example.

riyazhakki replied:
Thank you Subhasis for the reviewing the code, I will make the changes accordingly.

saimanikant commented:
[Ptl logs.txt](https://github.com/PBSPro/pbspro/files/2720382/Ptl.logs.txt)

saimanikant commented:
[ptl_log.txt](https://github.com/PBSPro/pbspro/files/2720658/ptl_log.txt)

anamikau commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -213,3 +220,39 @@ def test_execjob_end_non_blocking_multi_node(self):
+        if len(self.moms) > 1:
+            a = {'Resource_List.select': '2:ncpus=1',
+                 'Resource_List.place': "scatter"}
+            j = Job(TEST_USER, attrs=a)
```
I suggest optimizing the if-else condition by just initializing job object as that is only different. Rest can be common.

saimanikant replied:
Thanks! I've pushed a new commit addressing this.

anamikau commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -213,3 +220,39 @@ def test_execjob_end_non_blocking_multi_node(self):
+        """
+        hook_name = "execjob_end_logmsg6"
+        self.server.create_import_hook(hook_name, self.attr, self.hook_body)
+        if len(self.moms) > 1:
```
I think the check should be greater than 1 but less than2 since you are only requesting for 2 chunks.

vstumpf commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3080,343 @@ record_job_last_hook_executed(unsigned int hook_event,
+			}
+#endif
+		} else {
+			void post_reply(job *, int);
```
Why is this here? Shouldn't you include this from requests?

vstumpf commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3080,343 @@ record_job_last_hook_executed(unsigned int hook_event,
+				rpp_eom(server_stream); 
+
+			} else if (pjob->ji_qs.ji_svrflags & JOB_SVFLG_HERE) {
+#if MOM_ALPS
```
This code is now triplicated. here, 3229, and in req_deletejob. Maybe move it to a function?

vstumpf commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3080,343 @@ record_job_last_hook_executed(unsigned int hook_event,
+				}
+#endif
+			} else {
+				void post_reply(job *, int);
```
Again, why is this here?

vstumpf commented on src/resmom/mom_server.c:
```diff
@@ -1076,13 +1076,26 @@ is_request(int stream, int version)
+						if (pjob->ji_execjob_end_hook_event_started)
+							return;
 					(void)kill_job(pjob, SIGKILL);
+					phook_input = (mom_hook_input_t *) malloc (sizeof(mom_hook_input_t));
```
again, remove the spaces before and after malloc

vstumpf commented on src/resmom/requests.c:
```diff
@@ -619,23 +620,27 @@ req_deletejob(struct batch_request *preq)
 	 * prevent sending more OBIT messages
 	 */
+	pjob->ji_preq = preq;
+	hook_input = (mom_hook_input_t *) malloc (sizeof(mom_hook_input_t));
```
remove spaces around malloc

vstumpf commented on src/resmom/requests.c:
```diff
@@ -619,23 +620,27 @@ req_deletejob(struct batch_request *preq)
+			 * function 
+			 */
+			pjob->ji_execjob_end_hook_event_started = 1;
+			return;
```
Does this mean that the MS will run her hooks first, then only when she's done tell her sisters to run the hooks?

riyazhakki replied:
Yes and I see no difference in sequence after the code change.

vstumpf replied:
As far as I know, the reason why mother superior ran her hooks first, then the sisters, was because the hook would block the mom, preventing her from sending the IM_DELETE_JOB_REPLY requests to her sisters. Now that the hook doesn't block, shouldn't the request be sent earlier, before the hook finishes?

riyazhakki replied:
Ah, I overlooked this, thanks for pointing it out, I will address this in the next commit.

riyazhakki commented:
[ptl_log_multi_mom.txt](https://github.com/PBSPro/pbspro/files/2733145/ptl_log_multi_mom.txt)

riyazhakki commented:
Hi @vstumpf, Thanks for the review.
I have addressed your comments in the latest commit, have a look.

arungrover commented:
forget about my comments related to indentation, my github theme messed it up for me. you are fine

riyazhakki commented:
[ptl_log_multi_mom.txt](https://github.com/PBSPro/pbspro/files/2782415/ptl_log_multi_mom.txt)

riyazhakki commented:
Hi @prakashcv13 @arungrover @subhasisb @vstumpf @anamikau 

I have addressed your comments in the latest commits and also made changes to make the code generic for blocking/nonblocking mom hook.

Please review the code.

prakashcv13 commented on src/resmom/mom_hook_func.c:
```diff
@@ -664,7 +669,8 @@ static int
 	int parent_wait,  void (*post_func)(struct work_task *),
-	char *file_in,  char *file_out, char *file_data, size_t file_size)
+	char *file_in,  char *file_out, char *file_data, size_t file_size,
+	mom_process_hooks_params_t *php)
```
Add description for php in the doxygen.

prakashcv13 commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3088,437 @@ record_job_last_hook_executed(unsigned int hook_event,
+		log_class = PBS_EVENTCLASS_JOB;
+	}
+
+	mypid = php->child;
```
setting of mypid to php->child is not effective for background hooks as we set it to ptask->wt_event on line 3173.
Also, could you please explain how having the pid in the log will help?

riyazhakki replied:
We are forming the hook out file name using child's PID in the [post_run_hook](https://github.com/PBSPro/pbspro/pull/915/files#diff-f000d05f102c29b1985e578963befa7aR3217) function.
We are not using PID in logs in post_run_hook function.

prakashcv13 commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3088,437 @@ record_job_last_hook_executed(unsigned int hook_event,
+				job_purge(pjob);
+				dorestrict_user();
+
+				if ((ret=is_compose(server_stream, IS_DISCARD_DONE)) != DIS_SUCCESS)
```
spaces around '=' is needed

prakashcv13 commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3088,437 @@ record_job_last_hook_executed(unsigned int hook_event,
+				if ((ret=is_compose(server_stream, IS_DISCARD_DONE)) != DIS_SUCCESS)
+					goto err;
+
+				if ((ret=diswst(server_stream, jobid)) != DIS_SUCCESS)
```
spaces around '=' is needed

prakashcv13 commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3088,437 @@ record_job_last_hook_executed(unsigned int hook_event,
+				if ((ret=diswst(server_stream, jobid)) != DIS_SUCCESS)
+					goto err;
+
+				if ((ret=diswsi(server_stream, n)) != DIS_SUCCESS)
```
spaces around '=' is needed

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3153,9 +3589,23 @@ mom_process_hooks(unsigned int hook_event, char *req_user, char *req_host,
+	php->update_svr	= update_server;
+	php->hook_input	= hook_input;
+	php->hook_output	= hook_output;
+	php->parent_wait	= 1;
```
I do not see extra tabs on the last 2 lines.

riyazhakki commented:
Hi Prakash,
I have addressed your comments, please have a look.

vstumpf commented on src/resmom/mom_server.c:
```diff
@@ -1076,14 +1077,47 @@ is_request(int stream, int version)
-					mom_hook_input_init(&hook_input);
-					hook_input.pjob = pjob;
-					(void)mom_process_hooks(HOOK_EVENT_EXECJOB_END,
+					if ((phook_output->reject_errcode =
```
Why does `mom_hook_output_t->reject_errcode` need to be a pointer? Can't it just be an integer?

riyazhakki replied:
Existing declaration, is being used in other hook events too.

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3089,436 @@ record_job_last_hook_executed(unsigned int hook_event,
+		switch (wstat) {
+			case 0:
+				break;
+			case 254:
```
Yes, good to handle it now, made the changes.

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3089,436 @@ record_job_last_hook_executed(unsigned int hook_event,
+					phook->hook_name, pjob, hook_outfile);
+				hook_error_flag = 1;
+				break;
+			case 253:
```
Added

riyazhakki commented on src/resmom/mom_hook_func.c:
```diff
@@ -3075,6 +3089,436 @@ record_job_last_hook_executed(unsigned int hook_event,
+{
+	int	n = 0;
+	int	ret = 0;
+	char	jobid[PBS_MAXSVRJOBID+1] = {'0'};
```
Done

anamikau commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,261 @@
+        execjob_end hook with mutiple jobs
+        """
+        pltfom = self.du.get_platform()
+        if pltfom != 'cray':
```
not just cray but this won't work on a cpuset/cgroups enabled system as well. how about you just check if cpus is greater than 2 then run the test otherwise skip.

saimanikant replied:
Thanks! Addressed it in the new commit.

saimanikant commented:
test logs:
[TestPbsExecjobEnd.txt](https://github.com/PBSPro/pbspro/files/2840134/TestPbsExecjobEnd.txt)

vstumpf commented on test/tests/functional/pbs_hook_execjob_end.py:
```diff
@@ -0,0 +1,260 @@
+                                  'executed %s hook' % e.hook_name)\n"
+                     "e.accept()\n")
+        attr = {'event': 'execjob_end', 'order': '1', 'enabled': 'True'}
+        self.server.create_import_hook(hook_name1, attr, hook_body)
```
The variable names are inconsistent. `hook_name1` and `hook_body` are one hook, `hook_name` and `hook_body1` are another hook.
Maybe have `hook_body_reject` and `hook_body_accept`?

vstumpf commented on src/include/mom_func.h:
```diff
@@ -192,6 +192,9 @@ extern int   kill_job(job *, int sig);
 extern void  del_job_hw(job *);
 extern void  mom_deljob(job *);
+extern void  mom_deljob_wait2(job *);
+extern int  send_sisters_deljob_wait(job *);
```
All the other function names are lined up, please add a space.

riyazhakki commented:
Test logs 
[ptl_log_multi_mom.txt](https://github.com/PBSPro/pbspro/files/2850411/ptl_log_multi_mom.txt)

prakashcv13 commented on src/resmom/mom_hook_func.c:
```diff
@@ -3099,45 +3547,37 @@ record_job_last_hook_executed(unsigned int hook_event,
+	pbs_list_head		vnl_changes;
+	pbs_list_head		*head_ptr;
+	mom_process_hooks_params_t		*php = NULL;
+	struct work_task task;
```
the variable names need not be aligned, and in the above code, they are neither aligned nor placed after the type.

riyazhakki commented:
[ptl_log_multi_mom.txt](https://github.com/PBSPro/pbspro/files/2854427/ptl_log_multi_mom.txt)

prakashcv13 commented:
@riyazhakki - there are conflicts in mom_hook_func.c

riyazhakki commented:
I have now resolved conflicts and  rebased the branch with latest master.
Let me know if I can squash the commits

riyazhakki commented:
Attaching the test results and 
Here is the backup copy of missing commit 65e3c500de2f7c8a73a12c8c2cdbf742b0e4c63e of before the rebase.
[ptl_log_multi_mom.txt](https://github.com/PBSPro/pbspro/files/2856253/ptl_log_multi_mom.txt)

riyazhakki commented:
[ptl_log_multi_mom.txt](https://github.com/PBSPro/pbspro/files/2859072/ptl_log_multi_mom.txt)

prakashcv13 commented:
As codacy is acting up again, ignoring it and going ahead with the merge.

prakashcv13 commented:
merged #915
